packages/ra-core/src/CoreAdminRouter.tsx

```diff
@@ -7,53 +15,82 @@ import getContext from 'recompose/getContext';
 
 import { AUTH_GET_PERMISSIONS } from './auth/types';
 import { isLoggedIn } from './reducer';
-import { userLogout } from './actions/authActions';
+import { userLogout as userLogoutAction } from './actions/authActions';
 import RoutesWithLayout from './RoutesWithLayout';
-
-const welcomeStyles = {
+import {
+    Dispatch,
+    AuthProvider,
+    AdminChildren,
+    CustomRoutes,
+    CatchAllComponent,
+    LayoutComponent,
+    LayoutProps,
+} from './types';
+
+const welcomeStyles: CSSProperties = {
     width: '50%',
     margin: '40vh',
     textAlign: 'center',
 };
 
-export class CoreAdminRouter extends Component {
-    state = { children: [] };
+interface Props extends LayoutProps {
+    appLayout?: LayoutComponent;
+    catchAll?: CatchAllComponent;
+    children: AdminChildren;
+    customRoutes?: CustomRoutes;
+    loading?: ComponentType;
+}
+
+interface EnhancedProps {
+    authProvider?: AuthProvider;
+    isLoggedIn?: boolean;
+    userLogout: Dispatch<typeof userLogoutAction>;
+}
+
+interface State {
+    children: any[];
```

Comment:
why not use the AdminChildren type?

Reply:
Because the `children` property of the state is only used to store the result of the `children` **function** from the props. This function returns an array of `Resource` elements

Reply:
Actually we even know it's an array of Resource elements